src/lua/luaLoggerBinding.c

```diff
@@ -977,6 +979,257 @@ static int lua_update_gps(lua_State *L)
+
+        lua_validate_arg_number(L, 5);
+        uint8_t green = lua_tointeger(L, 5);
+
+        lua_validate_arg_number(L, 6);
+        uint8_t blue = lua_tointeger(L, 6);
+
+        lua_validate_arg_number(L, 7);
+        uint8_t flash = lua_tointeger(L, 7);
+
+        lua_pushinteger(L, shiftx_set_linear_threshold(threshold_id, segment_length, threshold, red, green, blue, flash));
+        return 1;
+}
+
+static int lua_sx_set_alert_threshold(lua_State *L)
+{
+        lua_validate_args_count(L, 7, 7);
+
+        lua_validate_arg_number(L, 1);
+        uint8_t alert_id = lua_tointeger(L, 1);
+
+        lua_validate_arg_number(L, 2);
+        uint8_t threshold_id = lua_tointeger(L, 2);
+
+        lua_validate_arg_number(L, 3);
+        uint16_t threshold = lua_tointeger(L, 3);
+
+        lua_validate_arg_number(L, 4);
+        uint8_t red = lua_tointeger(L, 4);
+
+        lua_validate_arg_number(L, 5);
+        uint8_t green = lua_tointeger(L, 5);
+
+        lua_validate_arg_number(L, 6);
+        uint8_t blue = lua_tointeger(L, 6);
+
+        lua_validate_arg_number(L, 7);
+        uint8_t flash = lua_tointeger(L, 7);
+
+        lua_pushinteger(L, shiftx_set_alert_threshold(alert_id, threshold_id, threshold, red, green, blue, flash));
+        return 1;
+}
+
+static int lua_sx_set_config(lua_State *L)
+{
+        lua_validate_args_count(L, 1, 5);
+
+        struct shiftx_configuration * shiftx_config = shiftx_get_config();
+
+        switch(lua_gettop(L)) {
+        default:
+                return lua_panic(L);
+        case 5:
+                /* auto brightness scaling */
+                lua_validate_arg_number(L, 8);
+                shiftx_config->auto_brightness_scaling = lua_tointeger(L, 8);
+        case 4:
+                /* CAN Base Address */
+                lua_validate_arg_number(L, 4);
+                shiftx_config->base_address = lua_tointeger(L, 4);
+        case 3:
+                /* CAN bus */
+                lua_validate_arg_number(L, 3);
+                shiftx_config->can_bus = lua_tointeger(L, 3);
+        case 2:
+                /* ShiftX brightness (0 to 100; 0=automatic brightness)*/
+                lua_validate_arg_number(L, 2);
+                shiftx_config->brightness = lua_tointeger(L, 2);
+        case 1:
+                /* ShiftX orientation (0=normal, 1= inverted) */
+                lua_validate_arg_number(L, 1);
+                shiftx_config->orientation_inverted = lua_tointeger(L, 1);
+        }
+
+        lua_pushinteger(L, shiftx_update_config());
+        return 1;
+}
+
+static int lua_rx_button_press(lua_State *L)
+{
+        uint8_t button_id;
+        uint8_t state;
+
+        if (shiftx_rx_button_press(&button_id, &state)) {
+                lua_pushinteger(L, button_id);
+                lua_pushinteger(L, state);
+                return 2;
+        }
+        else {
+                return 0;
+        }
+}
+
+static int lua_tx_button(lua_State *L)
```

Comment:
Should the second argument be a boolean and not an integer?  Either that or, assuming you're thinking of more than a bimodal state, should this be named tx_switch to be more generic?  Also, is the button id some opaque id that gets mapped on the app side or is it intended to map the the ascii keyboard event?  Either way it would be nice if a table could be pushed into the Lua state that maps logical button names to their ID value.

Reply:
Agreed, we can make it a boolean. Good catch!

Reply:
w/r/t using text labels for buttons, I don't think it's worth the complexity to map names at this point, given the embedded environment. We'll have a map of button codes that will correlate with CAN messages. I think it would be acceptable to expect users to remember 0=left button, 1=right button.

Reply:
I think that's fine, just needs to be addressed with documentation with a list of ids and their behavior.

Reply:
I like the idea of more than boolean for button state.  I use, quite often, multiple states such as { button_not_pressed, button_pressed, button_released, button_long_pressed, button_long_released }.  This allows the lower level code to differentiate short and long presses, as well as looking for change-of-state.  Say, for example, your code wants to look when a button is released, not just that it is not pressed.  Not-pressed is something more for start up whereas button released would indicate it was pressed then released.

Reply:
I was thinking more about rotary encoders when making the initial comment, like the idea of having an encoder to select a specific screen.  I'd have to look at what something like the Ecumaster CAN switch panel supports to know if the long/short press timing is calculated on the button panel.  I don't think ShiftX devices support that and since they're not field upgradeable it would have to be emulated in Lua.

Reply:
Doing a cursory look at the Ecumaster docs indicates that they only support binary state of the button, this would mean that if long/short press was to be implemented it would probably need to be on the RCP/Lua side.  The main use I could see for that is allowing a button to dual purpose as something like left/ok, right/cancel.

Reply:
We can also keep it 0/1 for now and expand it later to support more gestures (long press, double press, or whatever we come up with)